cpp/src/arrow/util/hashing.h

```diff
@@ -332,14 +332,24 @@ class HashTable {
 
 // XXX typedef memo_index_t int32_t ?
 
+// ----------------------------------------------------------------------
+// A base class for memoization table.
+
+class MemoTable {
+ public:
+  virtual ~MemoTable() = default;
+
+  virtual int32_t size() const = 0;
```

Comment:
Why is this needed?

Reply:
Because this change uses `size()` against `MemoTable` not subclass such as `ScalarMemoTable` here https://github.com/apache/arrow/pull/4255/files#diff-5562b5446398874441052ef04e7ee2e6R243 .
If this doesn't exist, we need to cast to subclass to call `size()`.